File: sdks/ios/Sources/AdxSDK/MmpIntegrationManager.swift
import Foundation

/// MMP Integration Manager for iOS
/// Supports: AppsFlyer, Adjust, Branch, Kochava, Singular
@MainActor
public class MmpIntegrationManager {
    
    public static let shared = MmpIntegrationManager()
    
    public enum MmpProvider: String {
        case appsflyer
        case adjust
        case branch
        case kochava
        case singular
        case tenjin
        case none
    }
    
    private var provider: MmpProvider = .none
    private var clickId: String?
    
    private init() {}
    
    // MARK: - AppsFlyer Integration
    
    /// Initialize AppsFlyer
    public func initializeAppsFlyer(devKey: String, appId: String) {
        #if canImport(AppsFlyerLib)
        import AppsFlyerLib
        
        provider = .appsflyer
        
        AppsFlyerLib.shared().appsFlyerDevKey = devKey
        AppsFlyerLib.shared().appleAppID = appId
        AppsFlyerLib.shared().delegate = self
        AppsFlyerLib.shared().customerUserID = AdxSDK.shared.userId
        
        // Start AppsFlyer
        AppsFlyerLib.shared().start()
        
        AdxSDK.shared.log("AppsFlyer initialized", level: .info)
        #else
        AdxSDK.shared.log("AppsFlyer SDK not found. Add AppsFlyerLib to your project.", level: .warning)
        #endif
    }
    
    // MARK: - Adjust Integration
    
    /// Initialize Adjust
    public func initializeAdjust(appToken: String, environment: String = "production") {
        #if canImport(Adjust)
        import Adjust
        
        provider = .adjust
        
        let adjustConfig = ADJConfig(
            appToken: appToken,
            environment: environment == "production" ? ADJEnvironmentProduction : ADJEnvironmentSandbox
        )
        
        adjustConfig?.logLevel = ADJLogLevelInfo
        
        // Attribution callback
        adjustConfig?.delegate = self
        
        Adjust.appDidLaunch(adjustConfig)
        
        AdxSDK.shared.log("Adjust initialized", level: .info)
        #else
        AdxSDK.shared.log("Adjust SDK not found. Add Adjust to your project.", level: .warning)
        #endif
    }
    
    // MARK: - Branch Integration
    
    /// Initialize Branch
    public func initializeBranch() {
        #if canImport(Branch)
        import Branch
        
        provider = .branch
        
        Branch.getInstance().initSession(launchOptions: nil) { params, error in
            if let error = error {
                AdxSDK.shared.log("Branch init error: \(error)", level: .error)
                return
            }
            
            if let params = params as? [String: Any] {
                self.handleBranchAttribution(params)
            }
        }
        
        AdxSDK.shared.log("Branch initialized", level: .info)
        #else
        AdxSDK.shared.log("Branch SDK not found. Add Branch to your project.", level: .warning)
        #endif
    }
    
    // MARK: - Click ID Management
    
    /// Set click ID from ad click
    public func setClickId(_ id: String) {
        clickId = id
        
        switch provider {
        case .appsflyer:
            #if canImport(AppsFlyerLib)
            AppsFlyerLib.shared().customData = [
                "adx_click_id": id,
                "adx_publisher_id": AdxSDK.shared.publisherId
            ]
            #endif
            
        case .adjust:
            #if canImport(Adjust)
            let event = ADJEvent(eventToken: "adx_click")
            event?.addCallbackParameter("click_id", value: id)
            event?.addCallbackParameter("publisher_id", value: AdxSDK.shared.publisherId)
            Adjust.trackEvent(event)
            #endif
            
        case .branch:
            #if canImport(Branch)
            Branch.getInstance().setRequestMetadataKey("adx_click_id", value: id)
            #endif
            
        default:
            break
        }
        
        AdxSDK.shared.log("Click ID set: \(id)")
    }
    
    /// Get current click ID
    public func getClickId() -> String? {
        return clickId
    }
    
    // MARK: - Event Tracking
    
    /// Track custom event
    public func trackEvent(
        eventName: String,
        eventValue: [String: Any]? = nil,
        revenue: Double? = nil,
        currency: String = "USD"
    ) {
        switch provider {
        case .appsflyer:
            #if canImport(AppsFlyerLib)
            AppsFlyerLib.shared().logEvent(eventName, withValues: eventValue)
            #endif
            
        case .adjust:
            #if canImport(Adjust)
            let event = ADJEvent(eventToken: eventName)
            eventValue?.forEach { key, value in
                event?.addCallbackParameter(key, value: "\(value)")
            }
            if let revenue = revenue {
                event?.setRevenue(revenue, currency: currency)
            }
            Adjust.trackEvent(event)
            #endif
            
        case .branch:
            #if canImport(Branch)
            let event = BranchEvent.customEvent(withName: eventName)
            event.customData = eventValue ?? [:]
            if let revenue = revenue {
                event.revenue = NSDecimalNumber(value: revenue)
                event.currency = .USD
            }
            event.logEvent()
            #endif
            
        default:
            // Track manually to taskirx
            Task {
                await trackEventTotaskirx(
                    eventName: eventName,
                    eventValue: eventValue,
                    revenue: revenue,
                    currency: currency
                )
            }
        }
    }
    
    // MARK: - Private Helpers
    
    private func handleBranchAttribution(_ params: [String: Any]) {
        if let clickId = params["+click_id"] as? String {
            setClickId(clickId)
            
            Task {
                await notifytaskirx(clickId: clickId, attributionData: params)
            }
        }
    }
    
    private func notifytaskirx(clickId: String, attributionData: [String: Any]) async {
        do {
            let endpoint = "\(AdxSDK.shared.getConfiguration().apiEndpoint)/api/mmp/attribution"
            
            let payload: [String: Any] = [
                "click_id": clickId,
                "publisher_id": AdxSDK.shared.publisherId,
                "mmp_provider": provider.rawValue,
                "attribution_data": attributionData,
                "device_id": AdxSDK.shared.userId,
                "install_time": Date().timeIntervalSince1970
            ]
            
            // Make HTTP request using NetworkClient
            // await AdxSDK.shared.getNetworkClient().post(endpoint, payload: payload)
            
            AdxSDK.shared.log("Attribution sent to platform for click_id: \(clickId)")
        } catch {
            AdxSDK.shared.log("Failed to send attribution: \(error)", level: .error)
        }
    }
    
    private func trackEventTotaskirx(
        eventName: String,
        eventValue: [String: Any]?,
        revenue: Double?,
        currency: String
    ) async {
        do {
            let endpoint = "\(AdxSDK.shared.getConfiguration().apiEndpoint)/api/mmp/event"
            
            let payload: [String: Any] = [
                "click_id": clickId as Any,
                "event_name": eventName,
                "event_value": eventValue as Any,
                "revenue": revenue as Any,
                "currency": currency,
                "timestamp": Date().timeIntervalSince1970
            ]
            
            // Make HTTP request
            // await AdxSDK.shared.getNetworkClient().post(endpoint, payload: payload)
            
            AdxSDK.shared.log("Event tracked: \(eventName)")
        } catch {
            AdxSDK.shared.log("Failed to track event: \(error)", level: .error)
        }
    }
}

// MARK: - AppsFlyer Delegate

#if canImport(AppsFlyerLib)
import AppsFlyerLib

extension MmpIntegrationManager: AppsFlyerLibDelegate {
    public func onConversionDataSuccess(_ conversionInfo: [AnyHashable : Any]) {
        guard let status = conversionInfo["af_status"] as? String else { return }
        
        if status == "Non-organic" {
            if let clickId = conversionInfo["af_sub1"] as? String {
                setClickId(clickId)
                
                Task {
                    await notifytaskirx(
                        clickId: clickId,
                        attributionData: conversionInfo as? [String: Any] ?? [:]
                    )
                }
            }
        }
    }
    
    public func onConversionDataFail(_ error: Error) {
        AdxSDK.shared.log("AppsFlyer conversion error: \(error)", level: .error)
    }
}
#endif

// MARK: - Adjust Delegate

#if canImport(Adjust)
import Adjust

extension MmpIntegrationManager: AdjustDelegate {
    public func adjustAttributionChanged(_ attribution: ADJAttribution?) {
        guard let attribution = attribution,
              let trackerToken = attribution.trackerToken else { return }
        
        setClickId(trackerToken)
        
        let attributionData: [String: Any] = [
            "tracker_name": attribution.trackerName ?? "",
            "tracker_token": trackerToken,
            "network": attribution.network ?? "",
            "campaign": attribution.campaign ?? "",
            "adgroup": attribution.adgroup ?? "",
            "creative": attribution.creative ?? ""
        ]
        
        Task {
            await notifytaskirx(clickId: trackerToken, attributionData: attributionData)
        }
    }
}
#endif

File: sdks/android/src/main/kotlin/com/adxplatform/sdk/mmp/MmpIntegrationManager.kt
package com.taskirx.sdk.mmp

import android.content.Context
import com.appsflyer.AppsFlyerLib
import com.appsflyer.attribution.AppsFlyerRequestListener
import com.adjust.sdk.Adjust
import com.adjust.sdk.AdjustConfig
import com.adjust.sdk.AdjustEvent
import com.adjust.sdk.LogLevel
import io.branch.referral.Branch
import com.taskirx.sdk.AdxSDK

/**
 * MMP Integration Manager for AdxSDK
 * Supports: AppsFlyer, Adjust, Branch, Kochava, Singular
 */
object MmpIntegrationManager {
    
    private var mmpProvider: MmpProvider? = null
    private var clickId: String? = null
    
    enum class MmpProvider {
        APPSFLYER,
        ADJUST,
        BRANCH,
        KOCHAVA,
        SINGULAR,
        TENJIN,
        NONE
    }
    
    /**
     * Initialize AppsFlyer
     */
    fun initializeAppsFlyer(
        context: Context,
        devKey: String,
        appId: String? = null
    ) {
        mmpProvider = MmpProvider.APPSFLYER
        
        AppsFlyerLib.getInstance().apply {
            init(devKey, null, context)
            start(context)
            
            // Set custom user ID from AdxSDK
            setCustomerUserId(AdxSDK.getUserId())
            
            // Register for attribution callback
            registerConversionListener(context, object : AppsFlyerConversionListener {
                override fun onConversionDataSuccess(data: Map<String, Any>?) {
                    data?.let {
                        val afStatus = it["af_status"] as? String
                        if (afStatus == "Non-organic") {
                            val clickId = it["af_sub1"] as? String
                            clickId?.let { id ->
                                setClickId(id)
                                notifytaskirx(id, it)
                            }
                        }
                    }
                }
                
                override fun onConversionDataFail(error: String?) {
                    // Handle error
                }
                
                override fun onAppOpenAttribution(data: Map<String, String>?) {
                    // Handle deep link
                }
                
                override fun onAttributionFailure(error: String?) {
                    // Handle error
                }
            })
        }
    }
    
    /**
     * Initialize Adjust
     */
    fun initializeAdjust(
        context: Context,
        appToken: String,
        environment: String = AdjustConfig.ENVIRONMENT_PRODUCTION
    ) {
        mmpProvider = MmpProvider.ADJUST
        
        val config = AdjustConfig(context, appToken, environment).apply {
            setLogLevel(LogLevel.INFO)
            
            // Set attribution callback
            setOnAttributionChangedListener { attribution ->
                attribution.trackerName?.let { trackerName ->
                    val clickId = attribution.trackerToken
                    clickId?.let { id ->
                        setClickId(id)
                        
                        val attributionData = mapOf(
                            "tracker_name" to trackerName,
                            "tracker_token" to attribution.trackerToken.orEmpty(),
                            "network" to attribution.network.orEmpty(),
                            "campaign" to attribution.campaign.orEmpty(),
                            "adgroup" to attribution.adgroup.orEmpty(),
                            "creative" to attribution.creative.orEmpty()
                        )
                        
                        notifytaskirx(id, attributionData)
                    }
                }
            }
        }
        
        Adjust.onCreate(config)
    }
    
    /**
     * Initialize Branch
     */
    fun initializeBranch(context: Context) {
        mmpProvider = MmpProvider.BRANCH
        
        Branch.getAutoInstance(context)
        
        // Branch initialization happens in Application.onCreate
        // Attribution will be received via Branch.sessionBuilder callback
    }
    
    /**
     * Set click ID from ad click
     */
    fun setClickId(id: String) {
        clickId = id
        
        // Pass to MMP if supported
        when (mmpProvider) {
            MmpProvider.APPSFLYER -> {
                AppsFlyerLib.getInstance().setAdditionalData(mapOf(
                    "adx_click_id" to id,
                    "adx_publisher_id" to AdxSDK.getPublisherId()
                ))
            }
            MmpProvider.ADJUST -> {
                // Adjust doesn't support setting click ID after init
                // But we can track it as an event
                val event = AdjustEvent("adx_click")
                event.addCallbackParameter("click_id", id)
                event.addCallbackParameter("publisher_id", AdxSDK.getPublisherId())
                Adjust.trackEvent(event)
            }
            MmpProvider.BRANCH -> {
                Branch.getInstance().setRequestMetadata("adx_click_id", id)
            }
            else -> {
                // Store for manual attribution
            }
        }
    }
    
    /**
     * Track custom event
     */
    fun trackEvent(
        eventName: String,
        eventValue: Map<String, Any>? = null,
        revenue: Double? = null,
        currency: String = "USD"
    ) {
        when (mmpProvider) {
            MmpProvider.APPSFLYER -> {
                AppsFlyerLib.getInstance().logEvent(
                    AdxSDK.getContext(),
                    eventName,
                    eventValue
                )
            }
            MmpProvider.ADJUST -> {
                val event = AdjustEvent(eventName)
                eventValue?.forEach { (key, value) ->
                    event.addCallbackParameter(key, value.toString())
                }
                revenue?.let {
                    event.setRevenue(it, currency)
                }
                Adjust.trackEvent(event)
            }
            MmpProvider.BRANCH -> {
                Branch.getInstance().userCompletedAction(eventName, eventValue)
            }
            else -> {
                // Track manually to taskirx
                trackEventTotaskirx(eventName, eventValue, revenue, currency)
            }
        }
    }
    
    /**
     * Get current click ID
     */
    fun getClickId(): String? = clickId
    
    /**
     * Get MMP provider
     */
    fun getMmpProvider(): MmpProvider? = mmpProvider
    
    /**
     * Notify taskirx about attribution
     */
    private fun notifytaskirx(clickId: String, attributionData: Map<String, Any>) {
        // Send attribution to taskirx backend
        kotlinx.coroutines.GlobalScope.launch {
            try {
                val endpoint = "${AdxSDK.getConfig().apiEndpoint}/api/mmp/attribution"
                val payload = mapOf(
                    "click_id" to clickId,
                    "publisher_id" to AdxSDK.getPublisherId(),
                    "mmp_provider" to mmpProvider?.name?.lowercase(),
                    "attribution_data" to attributionData,
                    "device_id" to AdxSDK.getUserId(),
                    "install_time" to System.currentTimeMillis()
                )
                
                // Make HTTP request (use your network client)
                // NetworkClient.post(endpoint, payload)
                
                println("AdxSDK: Attribution sent to platform for click_id: $clickId")
            } catch (e: Exception) {
                println("AdxSDK: Failed to send attribution: ${e.message}")
            }
        }
    }
    
    /**
     * Track event to taskirx
     */
    private fun trackEventTotaskirx(
        eventName: String,
        eventValue: Map<String, Any>?,
        revenue: Double?,
        currency: String
    ) {
        kotlinx.coroutines.GlobalScope.launch {
            try {
                val endpoint = "${AdxSDK.getConfig().apiEndpoint}/api/mmp/event"
                val payload = mapOf(
                    "click_id" to clickId,
                    "event_name" to eventName,
                    "event_value" to eventValue,
                    "revenue" to revenue,
                    "currency" to currency,
                    "timestamp" to System.currentTimeMillis()
                )
                
                // Make HTTP request
                // NetworkClient.post(endpoint, payload)
                
                println("AdxSDK: Event tracked: $eventName")
            } catch (e: Exception) {
                println("AdxSDK: Failed to track event: ${e.message}")
            }
        }
    }
}

/**
 * AppsFlyer Conversion Listener
 */
private interface AppsFlyerConversionListener : 
    com.appsflyer.AppsFlyerConversionListener {
    override fun onConversionDataSuccess(data: Map<String, Any>?)
    override fun onConversionDataFail(error: String?)
    override fun onAppOpenAttribution(data: Map<String, String>?)
    override fun onAttributionFailure(error: String?)
}
